### PR TITLE
Airspeed : new sensor SDP3X

### PR DESCRIPTION
**Pull request by rde-mato**
Hello,

This is a driver for the new sensor SDP3X,
It has been bench-tested with wind turbine on FMU-v2 (latest stable branch) and appears to work correctly.
I only had trouble with the parameter generation in Airspeed.cpp line 63.

Please review and let me know if I can help !

### CODE REVIEW

**Comment by magicrub** [comment]
this needs a rebase on master

File: libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        _dev->set_retries(10); 
+
+	// step 1 - reset on broadcast
+        uint8_t reset_cmd = SDP3X_RESET_CMD;
+        ret = _dev->transfer((uint8_t *)&reset_cmd, 1, nullptr, 0);
+        if (!ret)
+            printf("SDP3X: reset command failed\n");
+
+	// waiting 20ms to reset
+	usleep(20000);
```
Comment (magicrub) [inline libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp]: blocking is not allowed

**Comment by magicrub** [comment]
>I only had trouble with the parameter generation in Airspeed.cpp line 63

You don't need worry about this one, the param will generate on the wiki automatically once it has been merged into master.

File: libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        { 2, SDP3XD1_I2C_ADDR },
+        { 1, SDP3XD2_I2C_ADDR },
+        { 0, SDP3XD2_I2C_ADDR },
+        { 2, SDP3XD2_I2C_ADDR },
+    };
+    bool found = false;
+    bool ret = false;
+    
+    for (uint8_t i=0; i<ARRAY_SIZE(addresses); i++)
+    {
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp]: Brace should be on same line for *if*, *for*, *switch*, etc.

File: libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        if (!ret)
+            printf("SDP3X: get scale value failed\n");
+
+    	// Check the CRC
+        if (!_crc(&val[0], 2, val[2]) || !_crc(&val[3], 2, val[5]) || !_crc(&val[6], 2, val[8]))
+	    printf("SDP3X: CRC failed !\n");
+
+	_scale = (((uint16_t)val[6]) << 8) | val[7];
+	printf("Scale value is : %d\n", _scale);
+        _collect();
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp]: Indentation suddenly changed here.

File: libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+	ret = _dev->transfer(nullptr, 0, &val[0], sizeof(val));    
+        if (!ret)
+            printf("SDP3X: get scale value failed\n");
+
+    	// Check the CRC
+        if (!_crc(&val[0], 2, val[2]) || !_crc(&val[3], 2, val[5]) || !_crc(&val[6], 2, val[8]))
+	    printf("SDP3X: CRC failed !\n");
+
+	_scale = (((uint16_t)val[6]) << 8) | val[7];
+	printf("Scale value is : %d\n", _scale);
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp]: Is this really needed? There is a message next that is different depending on this value.

File: libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+// read the values from the sensor
+void AP_Airspeed_SDP3X::_collect()
+{
+	// read 9 bytes from the sensor
+	uint8_t val[6];
+	int ret = _dev->transfer(nullptr, 0, &val[0], sizeof(val));
+	if (!ret)
+		printf("SDP3X: Collect error\n");
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp]: Again, shouldn't this return immediately?

File: libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+	// Check the CRC
+	if (!_crc(&val[0], 2, val[2]) || !_crc(&val[3], 2, val[5]))
+	    printf("Error, BAD CRC collected\n");
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp]: Same.

File: libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+		_temp = temperature;
+	        sem->give();
+	}
+	_last_sample_time_ms = AP_HAL::millis();
+}
+
+// 20Hz timer
+void AP_Airspeed_SDP3X::_timer()
+{
+    _collect();
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp]: Nitpicking: if this method only calls `_collect`, why is it needed?

File: libraries/AP_Airspeed/AP_Airspeed_SDP3X.h [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+#define SDP3X_RESET_ADDR            	0x00
+#define SDP3X_RESET_CMD             	0x06
+#define SDP3X_CONT_MEAS_AVG_MODE    	0x3615
+
+#define SPD3X_MEAS_RATE			100
+#define SDP3X_MEAS_DRIVER_FILTER_FREQ	3.0f
+
+#define SDP3X_SCALE_PRESSURE_SDP31	60
+#define SDP3X_SCALE_PRESSURE_SDP32	240
+#define SDP3X_SCALE_PRESSURE_SDP33	20
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_SDP3X.h]: Please move the macros to the cpp file (and fix the indentation of two values while doing it).

File: libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+	float P_float = (float)P;
+	float temp_float = (float)temp;
+
+	float scale_float = (float)_scale;
+
+	float diff_press_pa = P_float / scale_float;
+	float temperature = temp_float / (float)SDP3X_SCALE_TEMPERATURE;
+
+        if (sem->take(HAL_SEMAPHORE_BLOCK_FOREVER))
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp]: I don't like taking forever, but I'll defer to @tridge on it. But the main issue is that the getters for pressure and temperature aren't using the semaphore.

File: libraries/AP_Airspeed/AP_Airspeed_SDP3X.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    float _get_temperature(int16_t dT_raw) const;
+    bool _crc(const uint8_t data[], unsigned size, uint8_t checksum);
+
+    float _temp;
+    float _press;
+    float _temperature;
+    float _pressure;
+    uint32_t _last_sample_time_ms;
+
+    uint16_t _scale{0};
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_SDP3X.h]: No need for this initialization, everything is zeroed on construction.

**Comment by WickedShell** [comment]
Closing in favor of #7174

**Pull request closed without merging**